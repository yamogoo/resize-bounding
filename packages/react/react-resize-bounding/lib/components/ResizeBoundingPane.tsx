import {
  useMemo,
  useRef,
  useState,
  MutableRefObject,
  type PropsWithChildren,
  type CSSProperties,
  useEffect,
} from "react";

import {
  PaneDirections,
  PaneDirectionAliases,
  type PaneDirectionKey,
  type IResizeBoundingClassNames,
  type IStyles,
  type Options,
} from "../shared/typings";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "./ResizeBoundingPane.styles";

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

export interface Props extends PropsWithChildren {
  prefix: string;
  direction: PaneDirectionKey;
  options: Options;
  classNames: IResizeBoundingClassNames;
  styles?: Partial<IStyles>;
  dragStart: (data: PaneEmittedData) => void;
  dragMove: (data: PaneEmittedData) => void;
  dragEnd: (data: PaneEmittedData) => void;
  focus: (state: boolean) => void;
}

const checkIsHorizontal = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(direction);

const checkIsVertical = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(direction);

const ResizeBoundingPane = ({
  children,
  classNames,
  direction,
  options,
  dragStart,
  dragMove,
  dragEnd,
  focus,
}: Props) => {
  const refPane: MutableRefObject<HTMLDivElement | null> = useRef(null);

  const [isFocused, setIsFocused] = useState(false);
  const [isPressed, setIsPressed] = useState(false);

  let isResizing = false;

  const paneComputedStyle = useMemo(() => {
    const _width = options?.width ?? 1;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;
      const _styles = paneBaseStyles(
        _width,
        _areaWidth,
        options?.position ?? "central",
      );
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }

    return;
  }, [options.width, options.activeAreaWidth, options?.position, direction]);

  const splitterComputedStyle = useMemo(() => {
    const _width = options?.width;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;

      const _styles = splitterBaseStyles(_width, _areaWidth);
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }
  }, [options.width, options.activeAreaWidth, direction]);

  const containerComputedStyles = useMemo(() => {
    const isHorizontal = checkIsHorizontal(direction);
    return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
  }, [direction]);

  const updateCursor = (state: boolean) => {
    let cursorActive: string;

    if (checkIsHorizontal(direction)) {
      cursorActive = options?.cursor?.horizontal ?? "auto";
    } else if (checkIsVertical(direction)) {
      cursorActive = options?.cursor?.vertical ?? "auto";
    } else cursorActive = "auto";

    if (refPane.current)
      refPane.current.style.cursor = state ? cursorActive : "auto";
  };

  /* * * Events * * */

  const emitFocus = (state: boolean) => {
    focus(state);
  };

  const onFocus = (e: PointerEvent, isFocused: boolean): void => {
    e.stopPropagation();
    onSelected(isFocused);
  };

  const onSelected = (state: boolean): void => {
    setIsFocused(state);
    updateCursor(state);
    emitFocus(state);
  };

  const onDragStart = (e: PointerEvent): void => {
    e.preventDefault();
    e.stopImmediatePropagation();

    if (!options.touchActions && e.pointerType === "touch") return;

    isResizing = true;
    setIsPressed(true);
    onSelected(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.setPointerCapture(e.pointerId);

    dragStart({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });

    const onDragMove = (e: PointerEvent): void => {
      dragMove({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    const onDragEnd = (e: PointerEvent): void => {
      isResizing = false;
      setIsPressed(false);

      const el = e.currentTarget as HTMLDivElement;
      el.releasePointerCapture(e.pointerId);

      el.removeEventListener("pointermove", onDragMove);
      el.removeEventListener("pointerup", onDragEnd);

      dragEnd({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    el.addEventListener("pointermove", onDragMove);
    el.addEventListener("pointerup", onDragEnd);
  };

  const onDragCancel = (e: PointerEvent): void => {
    const el = e.currentTarget as HTMLDivElement;

    isResizing = false;
    el.releasePointerCapture(e.pointerId);

    dragEnd({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });
  };

  const addEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.addEventListener("pointerenter", (e) => onFocus(e, true));
      el.addEventListener("pointerleave", (e) => onFocus(e, false));

      el.addEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  const removeEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.removeEventListener("pointerenter", (e) => onFocus(e, true));
      el.removeEventListener("pointerleave", (e) => onFocus(e, false));

      el.removeEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  useEffect(() => {
    addEventListeners();

    return () => {
      removeEventListeners();
    };
  });

  return (
    <div
      ref={refPane}
      data-testid="resize-bounding-pane"
      className={`${classNames.pane}${isPressed || isFocused ? " active" : ""}${options.addStateClasses ? ` ${isPressed ? "pressed" : isFocused ? "focused" : "normal"}` : ""}`}
      style={paneComputedStyle}
    >
      <div
        data-testid="resize-bounding-splitter"
        className={classNames.splitter}
        style={splitterComputedStyle}
      >
        {(isFocused || !options?.knob?.normalHidden) && options?.knob?.show && (
          <div
            data-testid="resize-bounding-splitter-container"
            className={classNames.splitterContainer}
            style={containerComputedStyles}
          >
            <div data-testid="resize-bounding-knob" className={classNames.knob}>
              {children}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default ResizeBoundingPane;
