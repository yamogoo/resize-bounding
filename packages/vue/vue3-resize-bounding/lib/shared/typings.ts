import fluentui from "@fluentui/merge-styles";

export interface Knob {
  show: boolean;
  normalHidden: boolean;
}

export enum SplitterPositions {
  CENTER = "central",
  EXTERNAL = "external",
  INTERNAL = "internal",
}

export type PanePosition = SplitterPositions | string;

export interface Options {
  prefix: string;
  width: number;
  activeAreaWidth?: number;
  position: PanePosition;
  cursor: Partial<
    Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
  >;
  knob: Partial<Knob>;
  touchActions: boolean;
}

export type ResizeBoundingClassNames =
  | "container"
  | "pane"
  | "splitter"
  | "splitterContainer"
  | "knob";

export interface IResizeBoundingClassNames {
  container: string;
  pane: string;
  splitter: string;
  splitterContainer: string;
  knob: string;
}

export interface IStyles extends Record<ResizeBoundingClassNames, IStyle> {}
export type IStyle = fluentui.IStyle;
export type IStyleSet = fluentui.IStyleSet;

/* * * Pane * * */

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}
