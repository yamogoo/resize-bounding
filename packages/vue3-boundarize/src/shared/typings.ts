import type { HTMLAttributes } from "vue";

export namespace BBResize {
  export interface Knob {
    show: boolean;
    constantlyShow: boolean;
  }

  export enum SplitterPositions {
    CENTER = "central",
    EXTERNAL = "external",
    INTERNAL = "internal",
  }

  export type PanePosition = SplitterPositions | string;

  export interface PaneOptions {
    width: number;
    position: PanePosition;
    showBorder: boolean;
    cursor: Partial<
      Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
    >;
    knob: Partial<Knob>;
  }

  export interface Options {
    prefix: string;
    pane: PaneOptions;
  }

  export interface StylingAttributes {
    class: string;
    style: HTMLAttributes["style"];
  }

  export type StylingStates = "focused";

  export type InteractiveElementStyles = Record<
    "splitter" | "knob",
    Record<StylingStates, StylingAttributes> & StylingAttributes
  >;

  export type PaneStyles = InteractiveElementStyles & StylingAttributes;

  export type Styles = Record<"container", StylingAttributes> & {
    pane: PaneStyles;
  };
}
