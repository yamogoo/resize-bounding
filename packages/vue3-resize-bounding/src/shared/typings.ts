import fluentui from "@fluentui/merge-styles";

export namespace ResizeBounding {
  export interface Knob {
    show: boolean;
    normalHidden: boolean;
  }

  export enum SplitterPositions {
    CENTER = "central",
    EXTERNAL = "external",
    INTERNAL = "internal",
  }

  export type PanePosition = SplitterPositions | string;

  export interface Options {
    prefix: string;
    width: number;
    position: PanePosition;
    cursor: Partial<
      Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
    >;
    knob: Partial<Knob>;
  }

  export type ResizeBoundingClassNames =
    | "container"
    | "pane"
    | "splitter"
    | "splitterContainer"
    | "knob";

  export interface IResizeBoundingClassNames {
    container: string;
    pane: string;
    splitter: string;
    splitterContainer: string;
    knob: string;
  }

  export interface IStyles extends Record<ResizeBoundingClassNames, IStyle> {}
  export type IStyle = fluentui.IStyle;
  export type IStyleSet = fluentui.IStyleSet;
}
