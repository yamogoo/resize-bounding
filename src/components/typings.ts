export interface BBResizeOptions {
  pane: {
    knob: {
      constantlyShow: boolean;
    };
  };
}

export interface BBResizeStyles {
  cursor: {
    active: {
      horizontal: string;
      vertical: string;
    };
  };
}
