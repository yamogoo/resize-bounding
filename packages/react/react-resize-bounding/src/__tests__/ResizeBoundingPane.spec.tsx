import { test, expect, describe, beforeEach, afterEach, vi } from "vitest";
import {
  createEvent,
  fireEvent,
  render,
  screen,
  cleanup,
} from "@testing-library/react";
import deepmerge from "deepmerge";
import "@testing-library/jest-dom";

import React from "react";

import { DataTestIds, PREFIX } from "./setup";
import { PaneDirectionAliases, PaneDirections } from "../../lib/shared/typings";

import ResizeBoundingPane, {
  PaneEmittedData,
  type Props,
} from "../../lib/components/ResizeBoundingPane";

import {
  defaultOptions,
  getClassNames,
} from "../../lib/components/ResizeBounding.classNames";
import {
  paneBaseStyles,
  splitterBaseStyles,
} from "../../lib/components/ResizeBoundingPane.styles";

const classNames = getClassNames({});

const requiredProps: Props = {
  prefix: PREFIX,
  direction: "r",
  options: defaultOptions,
  classNames,
  dragStart: vi.fn(),
  dragMove: vi.fn(),
  dragEnd: vi.fn(),
  focus: vi.fn(),
};

const getClassName = (el: Element, name: string): string | undefined => {
  return Object.values(el.classList).find((className) => className === name);
};

describe("ResizeBoundingPane", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();
    cleanup();
  });

  describe("classes", () => {
    describe("states", () => {
      test(`should have ".active" class name`, async () => {
        const props: Props = requiredProps;
        render(<ResizeBoundingPane {...props} />);

        const paneEl = screen.getByTestId(DataTestIds.PANE);
        const stateInit = getClassName(paneEl, "active");

        expect(!!stateInit).toBe(false);

        const ev = createEvent.pointerEnter(paneEl);
        fireEvent(paneEl, ev);

        const state = getClassName(paneEl, "active");

        expect(!!state).toBe(true);
        expect(state).toMatchSnapshot();
      });

      test('should not have ".normal" class name by default', () => {
        const props: Props = requiredProps;
        render(<ResizeBoundingPane {...props} />);
        const splitterEl = screen.getByTestId(DataTestIds.PANE);

        const className = getClassName(splitterEl, "normal");

        expect(!!className).toBeFalsy();
        expect(className).toMatchSnapshot();
      });

      const props: Props = deepmerge(requiredProps, {
        options: {
          addStateClasses: true,
        },
      });

      test('should have ".normal" class name by default (if addStateClasses)', () => {
        render(<ResizeBoundingPane {...props} />);
        const splitterEl = screen.getByTestId(DataTestIds.PANE);

        const className = getClassName(splitterEl, "normal");

        expect(!!className).toBeTruthy();
        expect(className).toMatchSnapshot();
      });

      test('should have ".focused" class name (if addStateClasses)', async () => {
        render(<ResizeBoundingPane {...props} />);
        const paneEl = screen.getByTestId(DataTestIds.PANE);

        const ev = createEvent.pointerEnter(paneEl);
        fireEvent(paneEl, ev);

        const className = getClassName(paneEl, "focused");

        expect(!!className).toBeTruthy();
        expect(className).toMatchSnapshot();
      });

      test('should have ".pressed" class name (if addStateClasses)', async () => {
        render(<ResizeBoundingPane {...props} />);
        const paneEl = screen.getByTestId(DataTestIds.PANE);

        const ev = createEvent.pointerDown(paneEl);
        fireEvent(paneEl, ev);

        const className = getClassName(paneEl, "pressed");

        expect(!!className).toBeTruthy();
        expect(className).toMatchSnapshot();
      });
    });
  });

  describe("events", () => {
    describe("focus", async () => {
      test(`should emit "focus" event with state (true)`, async () => {
        const props: Props = deepmerge(requiredProps, {
          direction: PaneDirections.RIGHT,
        });

        render(<ResizeBoundingPane {...props} />);
        const rootEl = screen.getByTestId(DataTestIds.PANE);

        const ev = createEvent.pointerEnter(rootEl);
        fireEvent(rootEl, ev);

        expect(props.focus).toHaveBeenCalled();
        expect(props.focus).toHaveBeenCalledWith(true);
        expect(props.focus).toMatchSnapshot();
      });

      test(`should emit "focus" with state (false)`, async () => {
        const props: Props = deepmerge(requiredProps, {
          direction: PaneDirections.RIGHT,
        });

        render(<ResizeBoundingPane {...props} />);

        const rootEl = screen.getByTestId(DataTestIds.PANE);

        const ev = createEvent.pointerLeave(rootEl);
        fireEvent(rootEl, ev);

        expect(props.focus).toHaveBeenCalled();
        expect(props.focus).toHaveBeenCalledWith(false);
        expect(props.focus).toMatchSnapshot();
      });
    });

    describe("drag", () => {
      const checkEvent = async (
        node: HTMLElement,
        eventName: string,
        eventData: { clientX: number; clientY: number },
        callback: (data: PaneEmittedData) => void,
      ) => {
        const ev = createEvent[eventName](node, eventData);
        fireEvent(node, ev);

        const x = eventData.clientX,
          y = eventData.clientY;

        expect(callback).toHaveBeenCalledWith({ dir: "r", x, y });
        expect(callback).toMatchSnapshot();
      };

      test(`should emit "dragStart -> dragMove -> dragEnd" with {x, y, dir}`, async () => {
        const props = deepmerge(requiredProps, {
          direction: PaneDirections.RIGHT,
        });

        render(<ResizeBoundingPane {...props} />, {});
        const rootEl = screen.getByTestId(DataTestIds.SPLITTER);

        checkEvent(
          rootEl,
          "pointerDown",
          { clientX: 1, clientY: 1 },
          props.dragStart,
        );

        checkEvent(
          rootEl,
          "pointerMove",
          { clientX: 2, clientY: 2 },
          props.dragMove,
        );

        checkEvent(
          rootEl,
          "pointerUp",
          { clientX: 3, clientY: 3 },
          props.dragEnd,
        );
      });
    });
  });

  describe("options", () => {
    describe("knob", () => {
      test("should always show knob", () => {
        const props = deepmerge(requiredProps, {
          options: {
            knob: {
              show: true,
            },
          },
        });

        render(<ResizeBoundingPane {...props} />);

        const knobEl = screen.getByTestId(DataTestIds.KNOB);

        expect(knobEl).toBeInTheDocument();
        expect(knobEl).toMatchSnapshot();
      });

      test("should not render the knob if { normalHidden: true }", () => {
        const props = deepmerge(requiredProps, {
          options: {
            knob: {
              show: true,
              normalHidden: true,
            },
          },
        });

        const { container } = render(<ResizeBoundingPane {...props} />);

        const knobEl = container.querySelector(
          `[data-testid="${DataTestIds.KNOB}"]`,
        );

        expect(knobEl).not.toBeInTheDocument();
        expect(knobEl).toMatchSnapshot();
      });

      test.each([
        [PaneDirections.LEFT, 90],
        [PaneDirections.RIGHT, 90],
        [PaneDirections.TOP, 0],
        [PaneDirections.BOTTOM, 0],
      ])("pane (--%s) should rotated to (%s deg)", (direction, deg) => {
        const props: Props = deepmerge(requiredProps, {
          direction,
          options: {
            knob: {
              show: true,
            },
          },
        });

        render(<ResizeBoundingPane {...props} />);

        const knobEl = screen.getByTestId(DataTestIds.SPLITTER_CONTAINER);
        const styles = knobEl.getAttribute("style");

        expect(styles).toContain(`(${deg}deg)`);
        expect(styles).toMatchSnapshot();
      });
    });
  });

  describe("styles", () => {
    describe("cursor", () => {
      test('default cursor style should be "auto"', async () => {
        const props: Props = deepmerge(requiredProps, {
          direction: PaneDirections.RIGHT,
        });

        render(<ResizeBoundingPane {...props} />);

        const rootEl = screen.getByTestId(DataTestIds.PANE);

        fireEvent.pointerEnter(rootEl);
        fireEvent.pointerLeave(rootEl);

        const cursorStyle = rootEl.getAttribute("style");

        expect(cursorStyle).toContain("auto");
        expect(cursorStyle).toMatchSnapshot();
      });

      const checkCursorActive = async (
        props: Props,
        comparedCursorStyle: string,
      ): Promise<void> => {
        render(<ResizeBoundingPane {...props} />);

        const rootEl = screen.getByTestId(DataTestIds.PANE);
        fireEvent.pointerEnter(rootEl);

        const cursorStyle = rootEl.getAttribute("style");

        expect(cursorStyle).toContain(comparedCursorStyle);
        expect(cursorStyle).toMatchSnapshot();
      };

      describe("active", () => {
        describe("default styles", () => {
          test.each([
            PaneDirectionAliases.HORIZONTAL,
            PaneDirections.RIGHT,
            PaneDirections.LEFT,
          ])(
            'default cursor style should be "col-resize" (--%s)',
            async (direction) => {
              const props = deepmerge(requiredProps, { direction });
              checkCursorActive(props, "col-resize");
            },
          );

          test.each([
            PaneDirectionAliases.VERTICAL,
            PaneDirections.TOP,
            PaneDirections.BOTTOM,
          ])(
            'default cursor style should be "row-resize" (--%s)',
            async (direction) => {
              const props = deepmerge(requiredProps, { direction });
              checkCursorActive(props, "row-resize");
            },
          );
        });

        describe("custom styles (options)", () => {
          const CUSTOM_ACTIVE_CURSOR = "ew-resize";

          test.each([...Object.values(PaneDirections)])(
            `custom cursor style should be "${CUSTOM_ACTIVE_CURSOR}" (--%s)`,
            async (direction) => {
              const props = deepmerge(requiredProps, {
                direction,
                options: {
                  cursor: {
                    vertical: CUSTOM_ACTIVE_CURSOR,
                    horizontal: CUSTOM_ACTIVE_CURSOR,
                  },
                  width: 0,
                  position: "central",
                  knob: {},
                  prefix: "",
                  touchActions: false,
                },
              });

              checkCursorActive(props, CUSTOM_ACTIVE_CURSOR);
            },
          );

          const SIZE = 12;
          const paneMap = paneBaseStyles(SIZE, SIZE, "central");
          const splitterMap = splitterBaseStyles(SIZE, SIZE);

          test.each([
            [PaneDirections.LEFT, paneMap.l],
            [PaneDirections.RIGHT, paneMap.r],
            [PaneDirections.TOP, paneMap.t],
            [PaneDirections.BOTTOM, paneMap.b],
          ])(
            `the pane (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const props = deepmerge(requiredProps, {
                direction,
                options: { width: SIZE },
              });

              render(<ResizeBoundingPane {...props} />);

              const paneEl = screen.getByTestId(DataTestIds.PANE);
              const style = paneEl.getAttribute("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );

          test.each([
            [PaneDirections.LEFT, splitterMap.l],
            [PaneDirections.RIGHT, splitterMap.r],
            [PaneDirections.TOP, splitterMap.t],
            [PaneDirections.BOTTOM, splitterMap.b],
          ])(
            `the pane splitter (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const props = deepmerge(requiredProps, {
                direction,
                options: { width: SIZE },
              });

              render(<ResizeBoundingPane {...props} />);

              const splitterEl = screen.getByTestId(DataTestIds.SPLITTER);
              const style = splitterEl.getAttribute("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );
        });
      });
    });

    describe("custom styles", () => {
      test.each([
        [{ left: "70px", right: "70px", width: "70px" }, PaneDirections.RIGHT],
      ])(
        "inner styles of splitter should not have to overrided by customStyles",
        async (_customStyles, direction) => {
          const SPLITTER_WIDTH = 12;

          const styles = splitterBaseStyles(SPLITTER_WIDTH, SPLITTER_WIDTH)[
            direction
          ];

          const props = deepmerge(requiredProps, {
            options: { width: SPLITTER_WIDTH },
          });

          render(<ResizeBoundingPane {...props} />);

          const splitterEl = screen.getByTestId(DataTestIds.SPLITTER);

          fireEvent.pointerEnter(splitterEl);
          const splitterStyles = splitterEl.getAttribute("style");

          if (styles && splitterStyles) {
            for (const [k, v] of Object.entries(styles)) {
              const value = `${k}: ${v.toString()}`.trim();
              const trimedStyles = splitterStyles.toString().trim();

              expect(trimedStyles).toContain(value);
              expect(value).toMatchSnapshot();
              expect(trimedStyles).toMatchSnapshot();
            }
            expect(splitterStyles).toMatchSnapshot();
          }
        },
      );
    });
  });

  describe("slots", () => {
    test.each([<p>knob</p>])(
      `should render knob slot (<p>knob</p>)`,
      (knobSlot) => {
        const props = deepmerge(requiredProps, {
          direction: "r",
          options: {
            knob: {
              show: true,
            },
          },
        });

        render(<ResizeBoundingPane {...props} children={knobSlot} />);

        const text = screen.getByTestId(DataTestIds.KNOB).textContent;
        expect(text).toBe("knob");
        expect(text).toMatchSnapshot();
      },
    );
  });
});
