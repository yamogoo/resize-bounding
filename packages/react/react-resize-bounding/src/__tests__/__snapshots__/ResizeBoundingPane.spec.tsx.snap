// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResizeBoundingPane > classes > states > should have ".active" class name 1`] = `"active"`;

exports[`ResizeBoundingPane > classes > states > should have ".focused" class name (if addStateClasses) 1`] = `"focused"`;

exports[`ResizeBoundingPane > classes > states > should have ".normal" class name by default (if addStateClasses) 1`] = `"normal"`;

exports[`ResizeBoundingPane > classes > states > should have ".pressed" class name (if addStateClasses) 1`] = `"pressed"`;

exports[`ResizeBoundingPane > classes > states > should not have ".normal" class name by default 1`] = `undefined`;

exports[`ResizeBoundingPane > events > drag > should emit "dragStart -> dragMove -> dragEnd" with {x, y, dir} 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "dir": "r",
        "x": 1,
        "y": 1,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResizeBoundingPane > events > drag > should emit "dragStart -> dragMove -> dragEnd" with {x, y, dir} 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "dir": "r",
        "x": 2,
        "y": 2,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResizeBoundingPane > events > drag > should emit "dragStart -> dragMove -> dragEnd" with {x, y, dir} 3`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "dir": "r",
        "x": 3,
        "y": 3,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResizeBoundingPane > events > focus > should emit "focus" event with state (true) 1`] = `
[MockFunction spy] {
  "calls": [
    [
      true,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResizeBoundingPane > events > focus > should emit "focus" with state (false) 1`] = `
[MockFunction spy] {
  "calls": [
    [
      false,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResizeBoundingPane > options > knob > pane (--b) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--l) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--r) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--t) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`ResizeBoundingPane > options > knob > should always show knob 1`] = `
<div
  class="resize-bounding-knob resize-bounding-knob-4"
  data-testid="resize-bounding-knob"
/>
`;

exports[`ResizeBoundingPane > options > knob > should not render the knob if { normalHidden: true } 1`] = `null`;

exports[`ResizeBoundingPane > slots > should render knob slot (<p>knob</p>) 1`] = `"knob"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--b) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--l) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--r) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--t) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--b) should have "{ left: '0px', bottom: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; bottom: -6px; width: 100%; height: 12px;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--l) should have "{ top: '0px', left: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; left: -6px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--r) should have "{ top: '0px', right: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; right: -6px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--t) should have "{ left: '0px', top: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; top: -6px; width: 100%; height: 12px;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--b) should have "{ top: '0px', height: '12px', width: '100%' }" styles 1`] = `"top: 0px; height: 12px; width: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--l) should have "{ right: '0px', width: '12px', height: '100%' }" styles 1`] = `"right: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--r) should have "{ left: '0px', width: '12px', height: '100%' }" styles 1`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--t) should have "{ bottom: '0px', height: '12px', width: '100%' }" styles 1`] = `"bottom: 0px; height: 12px; width: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "col-resize" (--h) 1`] = `"cursor: col-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "col-resize" (--l) 1`] = `"top: 0px; left: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "col-resize" (--r) 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "row-resize" (--b) 1`] = `"left: 0px; bottom: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "row-resize" (--t) 1`] = `"left: 0px; top: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "row-resize" (--v) 1`] = `"cursor: row-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > default cursor style should be "auto" 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: auto;"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 1`] = `"left: 0px"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 2`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 3`] = `"width: 12px"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 4`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 5`] = `"height: 100%"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 6`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 7`] = `"left: 0px; width: 12px; height: 100%;"`;
